pkg/sql/catalog/hydrateddesc/hydrate.go

```diff
@@ -25,7 +25,7 @@ import (
 
 // HydrationLookupFunc is the type of function required to look up type
 // descriptors and their parent schemas and databases when hydrating an object.
-type HydrationLookupFunc func(ctx context.Context, id descpb.ID) (catalog.Descriptor, error)
+type HydrationLookupFunc func(ctx context.Context, id descpb.ID, skipHydration bool) (catalog.Descriptor, error)
```

Comment:
I find that this added boolean argument makes `MakeTypeLookupFuncForHydration` really hard to reason about, especially now that it returns a recursive function. Instead can we make it so that the `HydrationLookupFunc` implementations _always_ set `SkipHydration` so that we _know_ that all `lookupFn` calls don't recurse further into the hydration logic?

This requires pulling some logic from the `hydrateDescriptors` method in `descs.Collection` into the recursive function returned by `MakeTypeLookupFuncForHydration` (specifically, after `typDesc, err = lookupFn(ctx, id, false /* skipHydration */)`), but that's probably a good thing anyway and will probably simplify code that's already complicated instead of complicating it further.

Reply:
I'll look into this in a bit.

Reply:
Sorry just came back to this...was looking at other issues. I think the right way to `hydrate a slice of descriptors` is that
(1). go through all target descriptors to figure out what types are need. this would just require getting all interested type OID/IDs. There this no look up needed. Simply just get all IDs. I don't think there'll be recursions. It's simply loop through each descriptor type internally to get types. Though there could be 2 or 3 levels. For example, for function, we'd need to go through args and return type, and if any of these type is a table, we need to go through all the visible table columns.

(2). then we construct a data structure which would contain all the types required with the IDs we get in (1). Here, we need to make sure all types are hydrated. Actually, `hydrating a type` seems the wrong concept here. Only table implicit record type needs hydration. So that, we could lookup all type descriptors first and memoize them in the new data structure. And then, we hydrate the table or implicit record type with the the type descriptors we already have. Now, we have all the types. 
-- There shouldn't be any recursion. Because:
    1> type descriptors doesn't need hydration, or 1 level hydration at most.
    2> table implicit record type only need 1 level hydration, because we can't use record type with in record type yet, and composite type can't use enum yet. Well, the worst case is that things can be very explicit.

(3) simply replace all the type references in each descriptor with the types we all ready got. there is not recursion.

Let me know what you think... But anyhow I don't think I can commit to this now. We can do this later.

Reply:
I think that you're completely correct. I'll be happy to do what you suggested as a follow-up to this PR.